addons/point_of_sale/static/src/app/store/db.js

```diff
@@ -567,22 +613,26 @@ export class PosDB {
         } catch {
             return [];
         }
-        var results = [];
-        for (var i = 0; i < this.limit; i++) {
-            var r = re.exec(this.category_search_string[category_id]);
-            if (r) {
-                var id = Number(r[1]);
-                const product = this.get_product_by_id(id);
-                if (!this.shouldAddProduct(product, results)) {
+        let results = [];
+        if(filteredProducts.length > 0) {
+            for (const filteredProduct of filteredProducts) {
+                let r = re.exec(filteredProduct.id + ":" + filteredProduct.display_name);
+                if(!this.add_match_product(r, results)) {
                     continue;
                 }
-                results.push(product);
-            } else {
-                break;
```

Comment:
There's no need to alter this section. Once you've identified the product IDs, simply update the `re` to exclusively encompass those IDs:
```js
const idsPattern = filteredProductIds.join('|');
const re = new RegExp(`(${idsPattern}):.*?` + unaccent(query), "gi");
```